File: app/src/main/java/mishka/mingeo/view/pumping/PumpingActivity.java
package mishka.mingeo.view.pumping;

import android.arch.lifecycle.Observer;
import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.design.widget.TabLayout;
import android.support.v4.view.ViewPager;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.support.v7.widget.Toolbar;
import android.view.Menu;
import android.view.MenuItem;

import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;
import butterknife.ButterKnife;
import mishka.mingeo.MinGeoApplication;
import mishka.mingeo.R;
import mishka.mingeo.data.model.Borehole;
import mishka.mingeo.data.model.Pumping;
import mishka.mingeo.di.component.DaggerPumpingActivityComponent;
import mishka.mingeo.di.component.PumpingActivityComponent;
import mishka.mingeo.di.module.PumpingModule;
import mishka.mingeo.view.BaseActivity;
import mishka.mingeo.view.BaseActivityKt;
import mishka.mingeo.view.pumping.pumpinginfo.BoreholeSummaryAdapter;
import mishka.mingeo.view.pumping.pumpingsummary.BoreholeSelectedListener;

public class PumpingActivity extends BaseActivityKt implements BoreholeSelectedListener {

    @BindView(R.id.borehole_pager)
    ViewPager boreholePager;

    @BindView(R.id.borehole_tabs)
    TabLayout tabs;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @Inject
    TabAdapter adapter;

    @Inject
    PumpingViewModel viewModel;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_pumping);

        ButterKnife.bind(this);

        Intent intent = getIntent();
        Pumping pumping = (Pumping) intent.getSerializableExtra("pumping");

        PumpingActivityComponent component = DaggerPumpingActivityComponent.builder()
                .applicationComponent(getApplicationComponent())
                .activityComponent(getActivityComponent())
                .pumpingModule(new PumpingModule(pumping))
                .build();

        component.inject(this);


        boreholePager.setAdapter(adapter);
        boreholePager.addOnPageChangeListener(adapter);
        viewModel.getBoreholes().observe(this, new Observer<List<Borehole>>() {
            @Override
            public void onChanged(@Nullable List<Borehole> boreholes) {
                adapter.updateTabs(boreholes);
            }
        });

        setSupportActionBar(toolbar);
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.pumping_menu, menu);
        return super.onCreateOptionsMenu(menu);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        if (item.getItemId() == R.id.add_borehole)
            onCreateBoreholeClick();
        return super.onOptionsItemSelected(item);
    }

    public void updateBoreholes(List<Borehole> boreholes) {
        adapter.updateTabs(boreholes);
    }

    public void onCreateBoreholeClick() {
        viewModel.onCreateBoreholeClick();
    }


    @Override
    public void onBoreholeSelected(int position) {
        boreholePager.setCurrentItem(position);
    }
}

File: app/src/main/java/mishka/mingeo/view/pumping/PumpingActivity.kt
package mishka.mingeo.view.pumping

import android.arch.lifecycle.Observer
import android.os.Bundle
import android.support.v7.widget.LinearLayoutManager
import android.view.View

import javax.inject.Inject

import kotlinx.android.synthetic.main.activity_pumping.*
import kotlinx.android.synthetic.main.property.view.*
import kotlinx.android.synthetic.main.toolbar.*
import mishka.mingeo.R
import mishka.mingeo.data.model.Borehole
import mishka.mingeo.data.model.Pumping
import mishka.mingeo.di.component.DaggerPumpingActivityComponent
import mishka.mingeo.di.component.PumpingActivityComponent
import mishka.mingeo.di.module.PumpingModule
import mishka.mingeo.view.BaseActivityKt
import mishka.mingeo.view.dialog.PumpPowerDialog
import mishka.mingeo.view.dialog.SetValueDialog
import mishka.mingeo.view.plot.PlotView
import mishka.mingeo.view.plot.PlotViewFragment
import mishka.mingeo.view.plot.PumpingPlotViewFragment
import mishka.mingeo.view.pumping.note.NotesFragment

class PumpingActivity : BaseActivityKt() {

    @Inject
    lateinit var viewModel: PumpingViewModel

    @Inject
    lateinit var adapter: BoreholeAdapter

    lateinit var component: PumpingActivityComponent

    private lateinit var notesFragment: NotesFragment

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_pumping)

        setSupportActionBar(toolbar)
        toolbar.title = "Выкачка"

        hideChart()

        val pumping = intent.getSerializableExtra("pumping") as Pumping

        component = DaggerPumpingActivityComponent.builder()
                .activityComponent(getActivityComponent())
                .pumpingModule(PumpingModule(pumping))
                .build()

        component.inject(this)
        component.inject(chartContainer as PlotViewFragment)
        (chartContainer as PumpingPlotViewFragment).pumping = pumping

        rvBoreholes.adapter = adapter
        rvBoreholes.layoutManager = LinearLayoutManager(this)

        viewModel.boreholes.observe(this, Observer {
            adapter.boreholes = it ?: ArrayList()
        })

        viewModel.getPumping().observe(this, Observer {
            pumpPower.propertyValue.text = it?.pumpPower.toString()
            adapter.centralBoreholeId = it?.centralBoreholeId
            if(it?.startPumpingTime != null){
                pumpingStartedLabel.visibility = View.VISIBLE
                btnStartPumping.visibility = View.GONE
            }

        })

        if(viewModel.isPumpingStarted()){
            pumpingStartedLabel.visibility = View.VISIBLE
        } else {
            btnStartPumping.visibility = View.VISIBLE
        }

        btnAddBorehole.setOnClickListener {
            onCreateBoreholeClick()
        }

        pumpPower.propertyName.text = "Мощность насоса"
        pumpPower.setOnClickListener { showSetPumpPowerDialog() }

        notesFragment = supportFragmentManager.findFragmentById(R.id.notesFragment) as NotesFragment

        btnStartPumping.setOnClickListener { onStartPumpingClick() }
    }

    private fun onStartPumpingClick() {
        viewModel.saveStartPumpingTime()
    }


    private fun hideChart() {
        chartContainer.view?.visibility = View.GONE
    }

    private fun showChart() {
        chartContainer.view?.visibility = View.VISIBLE
    }

    override fun onResume() {
        super.onResume()
        val data = viewModel.getSummaryDepth()

        for (boreholeDepths in data)
            if (boreholeDepths.isNotEmpty()) {
                showChart()
                break
            }

        (chartContainer as PlotView).update()
    }

    private fun showSetPumpPowerDialog() {
        PumpPowerDialog().apply {
            listener = object : SetValueDialog.SetValueListener {
                override fun onValueSet(value: Float) {
                    viewModel.setPumpPower(value)
                }
            }
        }.show(fragmentManager, "set_pump_power")
    }

    private fun onCreateBoreholeClick() = viewModel.onCreateBoreholeClick()

    fun markBoreholeAsCentral(borehole: Borehole?) {
        if (borehole != null)
            viewModel.setCentralBorehole(borehole)
    }


}
